test/utils.js

```diff
@@ -1,83 +1,118 @@
 'use strict'
 
-const ipfs = require('ipfs')
-const DaemonFactory = require('ipfsd-ctl')
-const delay = require('delay')
-
-const config = {
-  Addresses: {
-    API: '/ip4/0.0.0.0/tcp/0',
-    Gateway: '/ip4/0.0.0.0/tcp/0',
-    Swarm: ['/ip4/0.0.0.0/tcp/0', '/ip4/0.0.0.0/tcp/0/ws']
-  }
+const PeerId = require('peer-id')
+const PeerInfo = require('peer-info')
+const DuplexPair = require('it-pair/duplex')
+
+const Pubsub = require('libp2p-gossipsub')
+const { multicodec } = require('libp2p-gossipsub')
+
+const pWaitFor = require('p-wait-for')
+
+const createPeerInfo = async () => {
+  const peerId = await PeerId.create({ bits: 1024 })
+
+  return PeerInfo.create(peerId)
 }
 
-// spawn a daemon
-const spawnDaemon = () => {
-  const d = DaemonFactory.create({
-    exec: ipfs,
-    type: 'proc',
-    IpfsApi: require('ipfs-http-client')
-  })
+const createMockRegistrar = (registrarRecord) => ({
+  handle: () => {},
+  register: ({ multicodecs, _onConnect, _onDisconnect }) => {
+    const rec = registrarRecord[multicodecs[0]] || {}
 
-  return d.spawn({
-    disposable: true,
-    bits: 512,
-    config,
-    EXPERIMENTAL: {
-      pubsub: true
+    registrarRecord[multicodecs[0]] = {
+      ...rec,
+      onConnect: _onConnect,
+      onDisconnect: _onDisconnect
     }
-  })
+
+    return multicodecs[0]
+  },
+  unregister: () => {}
+})
+
+// as created by libp2p
+exports.createPubsubNode = async (registrarRecord) => {
+  const peerInfo = await createPeerInfo()
+  peerInfo.protocols.add(multicodec)
+  const pubsub = new Pubsub(peerInfo, createMockRegistrar(registrarRecord))
+
+  await pubsub.start()
+
+  return {
+    peerInfo: pubsub.peerInfo,
+    subscribe: (topic, handler) => {
+      pubsub.subscribe(topic)
+
+      pubsub.on(topic, handler)
+    },
+    unsubscribe: (topic, handler) => {
+      if (!handler) {
+        pubsub.removeAllListeners(topic)
+      } else {
+        pubsub.removeListener(topic, handler)
+      }
+
+      pubsub.unsubscribe(topic)
+    },
+    publish: (topic, data) => pubsub.publish(topic, data),
+    getTopics: () => pubsub.getTopics(),
+    getSubscribers: (topic) => pubsub.getSubscribers(topic),
+    stop: () => pubsub.stop()
+  }
```

Comment:
Personally I'd just return `pubsub` and refactor the tests that use `subscribe` and `unsubscribe` to use the libp2p PubSub API not the IPFS pubsub API. 😉

Reply:
These changes aim to use the libp2p pubsub API: https://github.com/libp2p/js-libp2p/blob/refactor/async-await/src/pubsub.js

My initial idea was to use the pubsub API directly. But, I would need to also change the `src` for doing the `subscribe` and `handle` in different operations. Also, for `js-ipfs`, I think it makes more sense to provide `libp2p.pubsub` to `datastore-pubsub`, instead of `libp2p.pubsub._pubsub` for having the API of the pubsub router directly.

I will go ahead and merge this, but we can discuss and make this change if you convince me that it is better 😛

Reply:
Ah gotcha my bad I was thinking the libp2p pubsub api WAS  the pubsub api but it is different.